### PR TITLE
[PIPELINES 4] Create release based on annotated git tag

### PR DESCRIPTION
**Pull request by petrkotas**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9739113/
Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9739185/
Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9741850/

### What this PR does / why we need it:

Changes to makefile:
- Introduces two new entrypoints in the `Makefile` that build and push tagged container image.
- Get tag only when it is annotated and attached to current commit, otherwise error.

Changed to pipelines:
- Add second build pipeline, 
	- chained after successful E2E
	- triggered when tag started the pipelines
	- Generates changelog from commits and summary provided in tag annotation
- Add second build image template 
	- build tagged image
	- needs to be added as a pipelines into the AZP

Signed-off-by: Petr Kotas <pkotas@redhat.com>

### Test plan for issue:

Manual. 

### Is there any documentation that needs to be updated for this PR?

Doc page once all bits are merged. The whole pipeline will get diagram and dedicated documentation page.

### CODE REVIEW

**Comment by petrkotas** [comment]
Contains commit from a https://github.com/Azure/ARO-RP/pull/1674 which need to go first.

File: .pipelines/build-and-push-images.yml [hunk 1]
```diff
@@ -5,8 +5,16 @@ pr: none
...
 
+resources:
+  pipelines:
+  - pipeline: buildImages
+    source: E2E
+    trigger: true
+
 jobs:
 - job: Build_and_push_images
+  condition: startsWith(variables['build.sourceBranch'], 'refs/tags/v')
```
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:16]: Unfortunately AZP does not support better filtering. No regex at this point.

File: .pipelines/build-and-push-images.yml [hunk 2]
```diff
@@ -5,8 +5,16 @@ pr: none
 variables:
 - template: vars.yml
 
+resources:
+  pipelines:
+  - pipeline: buildImages
+    source: E2E
```
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:11]: trigger after successful E2E run.

File: .pipelines/build-and-push-images.yml [hunk 3]
```diff
@@ -15,13 +23,30 @@ jobs:
...
       cp -a --parents aro "$(Build.ArtifactStagingDirectory)"
     displayName: Copy artifacts
   - task: PublishBuildArtifacts@1
     displayName: Publish Artifacts
     name: aro_deployer
+
+  - script: |
+      set -xe
+      MESSAGE="$(git for-each-ref refs/tags/$TAG --format='%(contents)')"
+      echo "$MESSAGE" > $(System.DefaultWorkingDirectory)/release.txt
```
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:43]: Extracts the annotation from the annotated tag. This is later used as input text for summary of the changelog.
The text have to be stored as a file as azp task can only load file.

File: .pipelines/build-and-push-images.yml [hunk 4]
```diff
@@ -15,13 +23,30 @@ jobs:
...
+    displayName: Extract release notes from annotated tag
+  - task: GithubRelease@0
```
Comment (petrkotas) [inline .pipelines/build-and-push-images.yml:45]: Release task generating the automated changelog in Github/Releases. It require service connection of type Github to Azure/ARO-RP granted by access token created at https://github.com/settings/tokens with write permission to the ARO-RP repository.

File: .pipelines/ci.yml [hunk 5]
```diff
@@ -3,6 +3,7 @@ trigger:
   branches:
     include:
     - master
+    - refs/tags/*
```
Comment (petrkotas) [inline .pipelines/ci.yml:6]: Trigger CI on any tag. Unfortunately no filtering available here.

File: .pipelines/deploy-azure-env.yml [hunk 6]
```diff
@@ -7,6 +7,13 @@ parameters:
 - name: vsoConfigBuildID
 - name: vsoDeployerBuildID
 
+resources:
```
Comment (petrkotas) [inline .pipelines/deploy-azure-env.yml:10]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/int-release.yml [hunk 7]
```diff
@@ -10,6 +10,13 @@ parameters:
   type: string
   default: latest
 
+resources:
```
Comment (petrkotas) [inline .pipelines/int-release.yml:13]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/prod-release.yml [hunk 8]
```diff
@@ -6,6 +6,13 @@ parameters:
 - name: vsoConfigBuildID
 - name: vsoDeployerBuildID
 
+resources:
```
Comment (petrkotas) [inline .pipelines/prod-release.yml:9]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/templates/template-job-deploy-azure-env.yml [hunk 9]
```diff
@@ -28,20 +28,7 @@ jobs:
       runOnce:
         deploy:
           steps:
-          - task: DownloadBuildArtifacts@0
```
Comment (petrkotas) [inline .pipelines/templates/template-job-deploy-azure-env.yml:31]: Part of https://github.com/Azure/ARO-RP/pull/1674

File: .pipelines/templates/template-push-images-to-acr.yml [hunk 10]
```diff
@@ -12,3 +12,14 @@ steps:
     export BRANCH=$(Build.SourceBranchName)
     make publish-image-aro
   displayName: ⚙️ Build and push images to ACR
+- script: |
```
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr.yml:15]: At this point this is direct copy of the previous step. First the steps are duplicated, once all works old can be deleted.

File: Makefile [hunk 11]
```diff
@@ -1,6 +1,8 @@
 SHELL = /bin/bash
+TAG = $(shell git describe)$(shell [[ $$(git status --porcelain) = "" ]] || echo -dirty)
```
Comment (petrkotas) [inline Makefile:2]: TAG is parsed by `git describe`, which returns nothing if there is no annotated tag.

**Comment by petrkotas** [comment]
E2E failed on tag push. Please do not rerun E2E. Build image got run because the trigger phase is not yet present in the pipelines and the code from this PR is ignored. Creating separate PR for trigger.

**Comment by petrkotas** [comment]
Converting to draft as I have figured out the path forward, which will require slight changes.

File: .pipelines/ci.yml [hunk 12]
```diff
@@ -6,6 +6,9 @@ trigger:
   paths:
     exclude:
     - docs/*
+  tags:
```
Comment (petrkotas) [inline .pipelines/ci.yml:9]: Added in https://github.com/Azure/ARO-RP/pull/1685

File: .pipelines/build-and-push-images-tagged.yml [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# Azure DevOps Pipeline building rp images and pushing to int acr
```
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:1]: Newly added pipeline chained after successful E2E pipeline when tag is the trigger. 

Requires adding as new pipeline to the AZP.

Duplicates the original pipeline, however it produces tagged image release. It also creates github release, which requires GitHub service connection.

File: .pipelines/build-and-push-images-tagged.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+      cp -a --parents aro "$(Build.ArtifactStagingDirectory)"
+    displayName: Copy artifacts
+  - task: PublishBuildArtifacts@1
+    displayName: Publish Artifacts
+    name: aro_deployer
+
+  - script: |
+      set -xe
+      MESSAGE="$(git for-each-ref refs/tags/$TAG --format='%(contents)')"
+      echo "$MESSAGE" > $(System.DefaultWorkingDirectory)/release.txt
```
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:43]: Extracts release summary from annotated tag. The summary is stored in the file, which is used as `releaseNotesFile` input for `GithubRelease@0` tasks. Task does not support other form of input.

Task also require `githubConnection`, which is service connection in ADO of type Github pointing to `ARO-RP` repository with write access.

Task also generates auto-changelog based on commit titles. To make it work nicely [conventional commits](https://conventionalcommits.org/en/v1.0.0/) are required.

File: .pipelines/deploy-azure-env.yml (hunk 6 above)
Comment (petrkotas) [inline .pipelines/deploy-azure-env.yml:10]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/int-release.yml (hunk 7 above)
Comment (petrkotas) [inline .pipelines/int-release.yml:13]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/prod-release.yml (hunk 8 above)
Comment (petrkotas) [inline .pipelines/prod-release.yml:9]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/templates/template-job-deploy-azure-env.yml [hunk 15]
```diff
@@ -28,20 +28,7 @@ jobs:
...
-              ${{ if eq(parameters.vsoConfigBuildID, 'latest') }}:
-                buildVersionToDownload: latestFromBranch
-                branchName: refs/heads/master
-              ${{ if ne(parameters.vsoConfigBuildID, 'latest') }}:
-                buildVersionToDownload: specific
-                buildId: ${{ parameters.vsoConfigBuildID }}
-              downloadType: specific
-              downloadPath: $(System.ArtifactsDirectory)/config
-            displayName: Download Config
+          - checkout: rpconfig
```
Comment (petrkotas) [inline .pipelines/templates/template-job-deploy-azure-env.yml:31]: Part of https://github.com/Azure/ARO-RP/pull/1674 which needs to go first.

File: .pipelines/templates/template-push-images-to-acr-tagged.yml [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+- script: |
+    set -e
+    trap 'set +e; for c in $(docker ps -aq); do docker rm -f $c; done; docker image prune -af ; rm -rf ~/.docker/config.json' EXIT
+
+    export RP_IMAGE_ACR=${{ parameters.rpImageACR }}
+
+    az acr login --name "$RP_IMAGE_ACR"
+    # azure checkouts commit, so removing master reference when publishing image
+    export BRANCH=$(Build.SourceBranchName)
+    make publish-image-aro-tag
```
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr-tagged.yml:13]: Copies the original pipeline, this is the only difference. Call tagged makefile entrypoint.

File: .pipelines/build-and-push-images-tagged.yml [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    name: ARO-CI
+
+  steps:
+  - template: ./templates/template-checkout.yml
+  - template: ./templates/template-az-cli-login.yml
+    parameters:
+      azureDevOpsJSONSPN: $(aro-v4-ci-devops-spn)
+
+
+  - template: ./templates/template-push-images-to-acr-tagged.yml
```
Comment (petrkotas) [inline .pipelines/build-and-push-images-tagged.yml:28]: Uses duplicated template with change to call tagged makefile entrypoint.

File: Makefile [hunk 18]
```diff
@@ -1,6 +1,8 @@
 SHELL = /bin/bash
+TAG = $(shell git describe --exact-match)
```
Comment (petrkotas) [inline Makefile:2]: Read the annotated tag from git. Requires the tag to be at current commit. Error otherwise.

File: Makefile [hunk 19]
```diff
@@ -50,6 +52,12 @@ image-aro: aro e2e.test
 	docker pull registry.access.redhat.com/ubi8/ubi-minimal
 	docker build --no-cache -f Dockerfile.aro -t $(ARO_IMAGE) .
 
+image-aro-tag: image-aro
+ifeq ($(TAG),)
```
Comment (petrkotas) [inline Makefile:56]: When no tag is present end. This happens when tag is not on current commit.

File: .pipelines/templates/template-push-images-to-acr-tagged.yml [hunk 20]
```diff
@@ -0,0 +1,15 @@
+parameters:
+  rpImageACR: ''
+  imageTag: ''
+steps:
+- script: |
+    set -e
+    trap 'set +e; for c in $(docker ps -aq); do docker rm -f $c; done; docker image prune -af ; rm -rf ~/.docker/config.json' EXIT
+
+    export RP_IMAGE_ACR=${{ parameters.rpImageACR }}
+    export TAG=${{ parameters.imageTag }}
```
Comment (petrkotas) [inline .pipelines/templates/template-push-images-to-acr-tagged.yml:10]: `TAG` is set here for visibility, it is also set in the `Makefile` for local work.

File: Makefile [hunk 21]
```diff
@@ -1,6 +1,8 @@
 SHELL = /bin/bash
+TAG ?= $(shell git describe --exact-match)
```
Comment (petrkotas) [inline Makefile:2]: Set variable when not already set

**Review by bryanro92 — APPROVED** [review]

**Review by bennerv — APPROVED** [review]
LGTM - wait for previous PRs to merge before this one.

**Review by s-amann — APPROVED** [review]

**Pull request merged**